libsolidity/formal/CHC.cpp

```diff
@@ -193,7 +211,17 @@ bool CHC::visit(FunctionDefinition const& _function)
 	auto functionPred = predicate(*functionEntryBlock, currentFunctionVariables());
 	auto bodyPred = predicate(*bodyBlock);
 
-	connectBlocks(m_currentBlock, functionPred);
+	if (_function.isConstructor())
+		connectBlocks(m_currentBlock, functionPred);
+	else
+		connectBlocks(genesis(), functionPred);
+
+	m_context.addAssertion(m_error.currentValue() == 0);
+	for (auto const* var: m_stateVariables)
+		m_context.addAssertion(m_context.variable(*var)->valueAtIndex(0) == currentValue(*var));
+	for (auto const& var: _function.parameters())
+		m_context.addAssertion(m_context.variable(*var)->valueAtIndex(0) == currentValue(*var));
```

Comment:
Just so I get this right: you could also instead use ``connectBlock(genesis(), initialFunctionVariables())`` above instead and have ``initialFunctionVariables()`` be the same as ``currentFunctionVariables()`` just with init variables being the same as the "mutable" variables (i.e. the current values, not the zero-index values)?
And you instead add the assertions that they are here, so that you don't need to basically duplicate what ``currentFunctionVariables()`` does?

Reply:
And in general the only reason for the "function entry block" to even have two sets of arguments is so that every block of the function deals with the same set of variables?

Reply:
At least for the arguments... for the state variables the "initial state" will always be required for supporting function calls and reverting I guess. Seems like a fine concept to me, just making sure I get it correctly.

Reply:
Yes, that's the idea. Connecting `genesis` with the variables directly wouldn't work because then its sort would also need to contains the variables of the functions.